torch/_prims/__init__.py

```diff
@@ -0,0 +1,566 @@
+      assert _dtype_to_kind(arg.dtype) is scalar_dtype, "Tensor with dtype " + str(arg.dtype) + " is not the expected dtype kind of " + str(scalar_dtype) + "!"
+    else:
+      assert False, "Unexpected type when checking for same dtype, " + str(type(arg)) + "!"
+
+#
+# Elementwise operations
+#
+
+# TODO: extend to handle scalar x scalar cases consistently
+
+def _elementwise_checks(*args, allow_scalars=True):
+  _check_same_device(*args, allow_scalars=allow_scalars)
+  _check_same_dtype(*args)
+  _check_same_shape(*args)
+
+def _make_elementwise_unary_op(op, *, allow_scalars=False):
+  def _op(a):
+    _elementwise_checks(a, allow_scalars=allow_scalars)
+    return op(a)
+  return _op
+
+def _make_elementwise_binary_op(op, *, allow_scalars=True):
+  def _op(a, b):
+    _elementwise_checks(a, b, allow_scalars=allow_scalars)
+    return op(a, b)
+  return _op
+
+
+#
+# Elementwise unary operations
+#
+
+abs = _make_elementwise_unary_op(torch.abs)
+
+# TODO: make sugar
+acos = _make_elementwise_unary_op(torch.acos)
+
+acosh = _make_elementwise_unary_op(torch.acosh)
+
+# TODO: make sugar
+asin = _make_elementwise_unary_op(torch.asin)
+
+# TODO: make sugar
+atan = _make_elementwise_unary_op(torch.atan)
+
+cos = _make_elementwise_unary_op(torch.cos)
+
+cosh = _make_elementwise_unary_op(torch.cosh)
+
+
+#
+# Elementwise binary operations
+#
+
+add = _make_elementwise_binary_op(torch.add)
+
+atan2 = _make_elementwise_binary_op(torch.atan2)
+
+div = _make_elementwise_binary_op(torch.true_divide)
+
+mul = _make_elementwise_binary_op(torch.mul)
+
+sub = _make_elementwise_binary_op(torch.sub)
+
+
+# View operations
+
+def _broadcast_in_dim_checks(
+  a: Tensor,
+  shape: Sequence[int],
+  broadcast_dimensions: Sequence[int]):
+    # Type checks
+    assert isinstance(a, Tensor)
+    assert isinstance(shape, Sequence)
+    assert isinstance(broadcast_dimensions, Sequence)
+
+    # every dimension must be accounted for
+    assert a.ndim == len(broadcast_dimensions)
+
+    # broadcast shape must have weakly more dimensions
+    assert len(shape) >= a.ndim
+
+    # broadcast_dimensions must be an ascending sequence
+    # (no relative reordering of dims) of integers and
+    # each dimension must be within the new shape
+    def _greater_than_reduce(acc, x):
+        assert isinstance(x, int)
+        assert x > acc
+        assert x < len(shape)
+
+        return x
+
+    reduce(lambda acc, x: _greater_than_reduce(acc, x), broadcast_dimensions, -1)
+
+    # shape must be broadcastable to
+    for idx, new_idx in enumerate(broadcast_dimensions):
+        assert a.shape[idx] == 1 or a.shape[idx] == shape[new_idx]
+
+def _broadcast_in_dim_aten(a, shape, broadcast_dimensions):
+    s = list(shape)
+    for broadcast_dimension in broadcast_dimensions:
+        s[broadcast_dimension] = -1
+
+    v = a
+    for idx, x in enumerate(s):
+        if x != - 1:
+            v.unsqueeze_(idx)
+
+    return v.expand(shape)
```

Comment:
Is this function equivalent to:
```python
def _broadcast_in_dim_aten(a, shape, broadcast_dimensions):
    idx_broad = 0
    v = a
    for idx in range(len(shape)):
        if idx != broadcast_dimensions[idx_broad]:
            v.unsqueeze_(idx)
        else:
            idx_broad += 1

    return v.expand(shape)
```
assuming that `broadcast_dimensions` is increasing?

Even after rewriting it, I don't understand what it does. Can you explain and leave a comment? Even leaving comments for some of the preconditions on the inputs would be very helpful (like `a.ndim == len(broadcast_dimensions)` and so.